### PR TITLE
Logic for checking Substrate proofs from within runtime module.

### PR DESCRIPTION
**Pull request by jimpo**
See PR title.

### CODE REVIEW

File: srml/bridge/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline srml/bridge/src/error.rs]:
```suggestion
// Copyright 2019 Parity Technologies (UK) Ltd.
```

File: srml/bridge/src/storage_proof.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline srml/bridge/src/storage_proof.rs]:
```suggestion
// Copyright 2019 Parity Technologies (UK) Ltd.
```

File: srml/bridge/src/lib.rs [hunk 3]
```diff
@@ -242,7 +246,7 @@ mod tests {
 				MockBridge::tracked_bridges(1),
 				Some(BridgeInfo {
 					last_finalized_block_number: 42,
-					last_finalized_block_hash: test_header.hash(), // FIXME: This is broken
```
Comment (kianenigma) [inline srml/bridge/src/lib.rs]: I don't know the details but is this now _unbroken_?

File: srml/bridge/src/storage_proof.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	db: MemoryDB<H>,
+}
+
+impl<H> StorageProofChecker<H>
+	where H: Hasher
+{
+	/// Constructs a new storage proof checker.
+	///
+	/// This returns an error if the given proof is invalid with respect to the given root.
+	pub fn new(root: H::Out, proof: Vec<Vec<u8>>) -> Result<Self, Error> {
```
Comment (kianenigma) [inline srml/bridge/src/storage_proof.rs]: This API does look a bit strange to me that you build a new instance of `checked` and via the constructor you also give it the _task_ that it has to do (i.e. proof to check). If it woks for you then it's okay but I was expecting `new` to actually return a new instance of something that _can check_ and a `.check` to actually check.

Comment (kianenigma) [reply srml/bridge/src/storage_proof.rs]: Though looking at the test I see it more clear of how you intend to use it. Just a thought.

File: srml/bridge/src/lib.rs (hunk 3 above)
Comment (jimpo) [reply srml/bridge/src/lib.rs]: Yes, this is now fixed in this PR. Was a compilation issue.

File: srml/bridge/src/storage_proof.rs (hunk 4 above)
Comment (jimpo) [reply srml/bridge/src/storage_proof.rs]: Yeah... I'm also not thrilled about this API, but it's kind of a direct consequence of the proof format. See https://github.com/paritytech/substrate/issues/3782.

`StateProofCheckerAndReader` is somewhat more accurate of a description, but in `substrate-state-machine`, it's just called `check_read_proof` usually so ¯\_(ツ)_/¯.

File: srml/bridge/src/lib.rs (hunk 3 above)
Comment (HCastano) [reply srml/bridge/src/lib.rs]: `testing::Header` didn't have a `hash()` method available, which is why it was broken. But yeah, it's fixed now

Comment (kianenigma) [reply srml/bridge/src/lib.rs]: My comment was due to the fact that only looking at the _local scope of the code_, I only identified that `test_header.hash();` now has a variable and no logic has changed and I was like _hmm.. how does that solve the problem?_ 😬

**Pull request merged**